### PR TITLE
feat: add Elixir macro annotation for tracing

### PR DESCRIPTION
**Pull request by btkostner**
This adds an `@trace` annotation for tracing functions. Code was mostly copied from the [New Relic Elixir agent](https://github.com/newrelic/elixir_agent/blob/master/lib/new_relic/tracer/macro.ex) also licensed under Apache License 2.0. I assume this is better code than I can come up with that's already used in production, but I'm willing to rewrite / re architect as needed.

I wasn't sure if / how you wanted tests considering there are none for the Elixir code.

### CODE REVIEW

File: apps/opentelemetry_api/lib/open_telemetry/macro.ex [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+  def traced_function_body(body, module, function, args, trace_info) do
+    {trace_name, trace_args} = build_trace_args(trace_info, module, function, args)
+
+    quote do
+      :otel_tracer.with_span(
+        :opentelemetry.get_application_tracer(__MODULE__),
+        unquote(trace_name),
+        Map.new(unquote(trace_args)),
+        fn _ -> unquote(body) end
```
Comment (btkostner) [inline apps/opentelemetry_api/lib/open_telemetry/macro.ex:169]: This is where the code is ran. I'm assuming (but want confirmation) that the `:otel_tracer.with_span` will capture any errors raised and attach them to the span.

**Comment by bryannaegele** [comment]
It's possible some time down the road. Also, I'm not sure if it's going to work with 1.0. Doesn't look like they've kept it up to date, but at first glance I don't think it will break.

**Comment by tsloughter** [comment]
Agh, looks like there are 2 https://hex.pm/packages/opentelemetry_decorator

Maybe we should have 3 :) and @btkostner you could send yours as a new app to https://github.com/open-telemetry/opentelemetry-erlang-contrib and we'll see if we can get a hold of `opentelemetry_decorator` on hex.

**Comment by connorlay** [comment]
I'm glad I found this issue! I've been working on an Elixir DSL for adding traces:

```elixir
def MyModule do

  use MyApp.OpenTelemetry

  trace :my_span_name, foo: "some custom attribute", bar: "another custom attribute"

  def my_traced_function() do
    # ...
  end

  trace :my_other_span_name
  
  defp my_private_traced_function() do
    # ...
  end
```

Why use a macro instead of a module attribute? So we can provide Elixir documentation for the DSL itself, plus macros are more flexible than module attributes when it comes to extending them and customizing their behavior. Currently my WIP DSL supports custom attributes and automatically records exceptions raised as a span error, before reraising to the caller.

I'm happy to share the work if folks are interested 🙂

**Pull request closed without merging**